src/Owin.Security.Providers.FreeConferenceCall/FreeConferenceCallAuthenticationHandler.cs

```diff
@@ -0,0 +1,257 @@
+                        if (primaryEmail != null)
+                        {
+                            context.Identity.AddClaim(new Claim(ClaimTypes.Email, primaryEmail.Email, XmlSchemaString,
+                        Options.AuthenticationType));
+                        }
+                    }
+                }
+                if (!string.IsNullOrEmpty(context.Name))
+                {
+                    context.Identity.AddClaim(new Claim("urn:freeconferencecall:name", context.Name, XmlSchemaString, Options.AuthenticationType));
+                }
+                if (!string.IsNullOrEmpty(context.Link))
+                {
+                    context.Identity.AddClaim(new Claim("urn:freeconferencecall:url", context.Link, XmlSchemaString, Options.AuthenticationType));
+                }
+                context.Properties = properties;
+
+                await Options.Provider.Authenticated(context);
+
+                return new AuthenticationTicket(context.Identity, context.Properties);
+            }
+            catch (Exception ex)
+            {
+                _logger.WriteError(ex.Message);
+            }
+            return new AuthenticationTicket(null, properties);
+        }
+
+        protected override Task ApplyResponseChallengeAsync()
+        {
+            if (Response.StatusCode != 401)
+            {
+                return Task.FromResult<object>(null);
+            }
+
+            var challenge = Helper.LookupChallenge(Options.AuthenticationType, Options.AuthenticationMode);
+
+            if (challenge == null) return Task.FromResult<object>(null);
+            var baseUri =
+                Request.Scheme +
+                Uri.SchemeDelimiter +
+                Request.Host +
+                Request.PathBase;
+
+            var currentUri =
+                baseUri +
+                Request.Path +
+                Request.QueryString;
+
+            var redirectUri =
+                baseUri +
+                Options.CallbackPath;
+
+            var properties = challenge.Properties;
+            if (string.IsNullOrEmpty(properties.RedirectUri))
+            {
+                properties.RedirectUri = currentUri;
+            }
+
+            // OAuth2 10.12 CSRF
+            GenerateCorrelationId(properties);
+
+            // comma separated
+            var scope = string.Join(",", Options.Scope);
+
+            var state = Options.StateDataFormat.Protect(properties);
+
+            var authorizationEndpoint =
+                Options.Endpoints.AuthorizationEndpoint +
+                "?client_id=" + Uri.EscapeDataString(Options.ClientId) +
+                "&redirect_uri=" + Uri.EscapeDataString(redirectUri) +
+//                "&scope=" + Uri.EscapeDataString(scope) +
+//                "&state=" + Uri.EscapeDataString(state) +
+                "&response_type=" + Uri.EscapeDataString("token"); // or "code"
```

Comment:
Consider making this a parameter/option (and remove the commented out scope and state)